// siml-angular-brunch/lib/index.js
// Generated by CoffeeScript 1.6.2
var SIMLCompiler, fs, mkdirp, siml, sysPath, write;

fs = require('fs');

siml = require('siml');

sysPath = require('path');

mkdirp = require('mkdirp');

write = function(path, content, append) {
  var dir;

  if (append == null) {
    append = false;
  }
  if (content == null) {
    return;
  }
  dir = sysPath.dirname(sysPath.normalize(path));
  return mkdirp(dir, '0775', function(err) {
    if (err != null) {
      throw err;
    }
    if (append != null) {
      return fs.appendFile(path, content, function(err) {
        if (err != null) {
          throw err;
        }
      });
    } else {
      return fs.writeFile(path, content, function(err) {
        if (err != null) {
          throw err;
        }
      });
    }
  });
};

module.exports = SIMLCompiler = (function() {
  SIMLCompiler.prototype.brunchPlugin = true;

  SIMLCompiler.prototype.type = 'template';

  SIMLCompiler.prototype.extension = 'siml';

  function SIMLCompiler(config) {
    var _ref, _ref1, _ref2, _ref3, _ref4, _ref5;

    this.templates = [];
    this["public"] = config.paths["public"];
    this.outFile = (Object.keys(config.files.templates.joinTo))[0] || 'templates.js';
    this.rootDir = config.files.templates.joinTo[this.outFile];
    this.createRouter = !!((_ref = config.plugins) != null ? (_ref1 = _ref.siml) != null ? _ref1.createRouter : void 0 : void 0);
    this.routerOptions = (_ref2 = config.plugins) != null ? (_ref3 = _ref2.siml) != null ? _ref3.routerOptions : void 0 : void 0;
    this.templateModuleName = ((_ref4 = config.plugins) != null ? (_ref5 = _ref4.siml) != null ? _ref5.moduleName : void 0 : void 0) || 'templates';
  }

  SIMLCompiler.prototype.compile = function(data, path, callback) {
    var content, e, error, name, path_hunks;

    try {
      content = siml.angular.parse(data, {
        pretty: false
      });
      path = path.replace(this.rootDir, '');
      path_hunks = path.split(sysPath.sep);
      name = path_hunks.pop().slice(0, -this.extension.length - 1);
      path_hunks.push(name);
      return this.templates.push({
        name: name,
        content: content,
        path: sysPath.join.apply(this, path_hunks)
      });
    } catch (_error) {
      e = _error;
      error = "Error: " + e.message;
      if (e.type) {
        error = e.type + error;
      }
      if (e.filename) {
        return error += " in '" + e.filename + ":" + e.line + ":" + e.column + "'";
      }
    } finally {
      callback(error, '');
    }
  };

  SIMLCompiler.prototype.onCompile = function(generated_files) {
    var router_module, template_module;

    router_module = this.createRouter ? '\n' + this.getRouterModule(this.templates) || '' : '';
    template_module = this.getTemplateModule(this.templates);
    return write("" + this["public"] + sysPath.sep + this.outFile, template_module + router_module, true);
  };

  SIMLCompiler.prototype.getTemplateModule = function(templates) {
    var content, escaped_content, template, _i, _len;

    content = '';
    for (_i = 0, _len = templates.length; _i < _len; _i++) {
      template = templates[_i];
      escaped_content = template.content.replace(/'/g, "\\'");
      content += "\n    $templateCache.put('" + template.path + "', '" + escaped_content + "');";
    }
    return "angular.module('" + this.templateModuleName + "', [])\n  .run(['$templateCache', function($templateCache) {\n    " + content + "\n  }])";
  };

  SIMLCompiler.prototype.getRouterModule = function(templates) {
    var content, controller_name, route_name, template, _i, _len;

    content = '';
    for (_i = 0, _len = templates.length; _i < _len; _i++) {
      template = templates[_i];
      if (!((this.routerOptions.onlyUse != null) && template.path.indexOf(this.routerOptions.onlyUse) === 0)) {
        continue;
      }
      route_name = template.path.slice(template.path.indexOf(sysPath.sep, 1)).slice(0, -5);
      controller_name = template.name[0].toUpperCase() + template.name.slice(1);
      content += "\n    $routeProvider.when('" + route_name + "', { controller: '" + controller_name + "', templateUrl: '" + template.path + "' });";
    }
    if (this.routerOptions.defaultRoute != null) {
      content += "\n    $routeProvider.otherwise({ redirectTo: '" + this.routerOptions.defaultRoute + "' });";
    }
    return "angular.module('" + (this.routerOptions.moduleName || 'router') + "', [])\n  .config(['$routeProvider', function($routeProvider) {\n    " + content + "\n  }])";
  };

  return SIMLCompiler;

})();
